pybossa/exporter/csv_reports_export.py

```diff
@@ -41,69 +40,60 @@ def download_name(self, project, ty):
                            'First Task Submission', 'Last Task Submission', 'Average Time Per Task']
-            writer.writerow(user_section)
 
             # get user data for report
-            users_project_data = get_project_report_userdata(id, start_date, end_date)
-            if users_project_data:
-                writer.writerow(user_header)
-                for user_data in users_project_data:
-                    writer.writerow(user_data)
+            users_project_data = get_project_report_userdata(project_id, start_date, end_date)
+            if users_project_data and users_project_data[0]:
+                users_csv = pd.DataFrame(users_project_data, columns=user_header).to_csv(index=False)
             else:
-                writer.writerow(['No user data'])
+                users_csv = pd.DataFrame(columns=user_header).to_csv(index=False)
+
+            csv_txt = f'{project_section}\n{project_csv}\n{user_section}\n{users_csv}'
 
-            return self._get_csv(out, writer)
+            return csv_txt
 
     def _make_zip(self, project, ty, **kwargs):
         name = self._project_name_latin_encoded(project)
-        csv_task_generator = self._respond_csv(ty, project.id, **kwargs)
+        csv_task_generator = self._respond_csv(ty, project.id)
         if csv_task_generator is not None:
-            with tempfile.NamedTemporaryFile() as datafile, \
-                 tempfile.NamedTemporaryFile(delete=False) as zipped_datafile:
+            datafile = tempfile.NamedTemporaryFile()
+            try:
+                datafile.write(str(csv_task_generator).encode('utf-8'))
+                datafile.flush()
+                zipped_datafile = tempfile.NamedTemporaryFile()
                 try:
-                    for line in csv_task_generator:
-                        datafile.write(str(line).encode())  # write() accepts bytes only
-                    datafile.flush()
-                    csv_task_generator.close()  # delete temp csv file
                     _zip = self._zip_factory(zipped_datafile.name)
                     _zip.write(
                         datafile.name,
                         secure_filename('%s_%s.csv' % (name, ty)))
                     _zip.close()
-                    return dict(filepath=zipped_datafile.name,
-                                filename=self.download_name(project, ty),
-                                delete=True)
-                except Exception:
-                    if os.path.exists(zipped_datafile.name):
-                        os.remove(zipped_datafile.name)
-                    raise
+                    container = "user_%d" % project.owner_id
+                    _file = FileStorage(
+                        filename=self.download_name(project, ty),
+                        stream=zipped_datafile)
+                    uploader.upload_file(_file, container=container)
+                finally:
```

Comment:
in case of exception, temp files won't be cleaned up?

Reply:
It will; ```zipped_datafile.close()``` and ```datafile.close()``` in the finally block will do the trick because NamedTemporaryFile has the parameter ```delete``` default to True. And it will delete the file when closing. Refer [here](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile).